csrc/multidevice/utils.cpp

```diff
@@ -151,6 +217,106 @@ void insertReshardings(Fusion* fusion) {
   }
 }
 
+void insertShardedAxisReordering(Fusion* fusion) {
+  auto exprs = fusion->exprs();
+  std::vector<Expr*> reshard_exprs;
+  for (auto expr : exprs) {
+    if (isResharding(expr)) {
+      reshard_exprs.push_back(expr);
+    }
+  }
+  for (auto expr : reshard_exprs) {
+    NVF_ERROR(
+        ir_utils::isTvOp(expr),
+        "Non-tv op is not supported yet: ",
+        expr->toString());
+    NVF_ERROR(
+        expr->outputs().size() == 1,
+        "Resharding operations can only have one output");
+    NVF_ERROR(
+        expr->inputs().size() == 1,
+        "Resharding operations can have only one input");
+    auto output = expr->outputs().at(0)->as<TensorView>();
+    auto input = expr->inputs().at(0)->as<TensorView>();
+    auto [shard_additions, shard_deletions] = allocationShardings(expr);
+    NVF_ERROR(
+        shard_additions.size() + shard_deletions.size() <= 1,
+        "Resharding expr can only support one axis")
+
+    // For gather operations i.e. ID goes from sharded to unsharded
+    // this will rematerialize a sharded axis.
+    // ProcessGroup expects contiguous tensors.
+    // Update input to push the rematerialized axis to the front -> collective
+    // -> permute the rematerizlied axis to the proper location Example: [i0
+    // DIDx(i1)] -> [i0 i1] Rewritten to: [DIDx(i1) i0] -> [i1 i0] -> [i0 i1]
+    // Note: there are no reduction based collectives that materializes an axis
+    // so expr is guaranteed to be a set.
+    if (!shard_deletions.empty()) {
+      auto id = shard_deletions[0];
+      int idx = static_cast<int>(input->domain()->posOf(id));
+      if (isOutermostAllocatedId(input, id)) {
+        continue;
+      }
+      TensorView* input_permute = permute(input, {{idx, 0}});
+      TensorView* output_permute = set(input_permute);
+      TensorView* new_output = permute(output_permute, {{0, idx}});
+      ir_utils::replaceValInAllExprInputsAndFusionOutputs(output, new_output);
+      for (int i = 0; i < static_cast<int>(output->nDims()); i++) {
+        new_output->axis(i)->parallelize(output->axis(i)->getParallelType());
+      }
+      new_output->setDeviceMesh(output->getDeviceMesh());
+      shardAllLike(input, {input_permute});
+      shardAllLike(new_output, {output_permute});
+    }
+    // For scatter operations i.e. ID goes from unsharded to sharded
+    // Update input to push the scattered axis to the front -> collective ->
+    // permute the sharded axis to the proper location.
+    // Example: [i0 i1 DIDx(i2)] -> [i0 DIDx(i1) r2]
+    // Rewritten to: [i0 i1 DIDx(i2)] -> [i1 i0 DIDx(i2)] ->
+    //                    [DIDx(i1) i0 r2] -> [i0 DIDx(i1)]
+    else if (!shard_additions.empty()) {
+      auto id = shard_additions[0];
+      int idx = static_cast<int>(output->domain()->posOf(id));
```

Comment:
```suggestion
      int sharding_axis = static_cast<int>(output->domain()->posOf(id));
```

I believe `axis` and `idx` here are used interchangeably, but better be consistent. I added `sharding_` because code below also talks about reduction axis, and `sharding_`  makes the name less ambiguous.